src/structural_properties.c

```diff
@@ -6897,3 +6897,185 @@ int igraph_i_is_graphical_degree_sequence_directed(
 
 #undef SUCCEED
 #undef FAIL
+
+
+/* igraph_is_tree -- check if a graph is a tree */
+
+/* count the number of vertices reachable from the root */
+static int igraph_i_is_tree_visitor(igraph_integer_t root, const igraph_adjlist_t *al, igraph_integer_t *visited_count) {
+    igraph_stack_int_t stack;
+    igraph_vector_bool_t visited;
+    long i;
+
+    IGRAPH_CHECK(igraph_vector_bool_init(&visited, igraph_adjlist_size(al)));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &visited);
+
+    IGRAPH_CHECK(igraph_stack_int_init(&stack, 0));
+    IGRAPH_FINALLY(igraph_stack_int_destroy, &stack);
+
+    *visited_count = 0;
+
+    /* push the root into the stack */
+    igraph_stack_int_push(&stack, root);
```

Comment:
Pushing to the stack should be protected with `IGRAPH_CHECK()` as we may run out of memory here.